### PR TITLE
Tina Content API - use .ts files for Tina Cloud config

### PR DESCRIPTION
**Pull request by jeffsee55**
[View file](https://github.com/tinacms/rfcs/blob/495d65ca58d0b518bd00cc51023bbfef77c73e43/0014-tina-config-ts.md)

### CODE REVIEW

File: 0014-tina-config-ts.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+    type: text # we know the expected type from this
+```
+
+In Tina, there's no concept of a `type` field, instead you have a `component` field, which offers no insight about the type of field. Strong types are essential to the Content API working properly, so we'd need to add an additional field to the traditional Tina form config. We could pass the `component` value through if it's provided, otherwise fall back to the `type`.
+
+## Caveats
+
+### Runtime code
+
+We don't have a mechanism (right now) to pull in these definitions to the runtime Tina form, so things like `parse`,`format` and `validate` would be a no-op for this stage of the feature. And `component`, which can be either a string or a React component, would need to be a string.
```
Comment (jamespohalloran) [inline 0014-tina-config-ts.md:68]: I like how this opens up the door for custom functions in the future, but I think it's fine that these are a no-op for now.

**Review by jamespohalloran — COMMENTED** [review]
I like this idea a lot.
- I think devs will be much more comfortable writing js/ts/json. 
- The runtime section sounds like it has cool future possibilities with Tina
- I think this change will help us shed some of the forestry-config lingo in the process. (E.g Would now be a good time to consider naming changes like ".tina/templates/front_matter" -> ".tina/models"? )

**Comment by jeffsee55** [comment]
@jamespohalloran yeah I think we could go ahead and remove Forestry-isms in this process. The compilation step would add them back under the hood, but just as a stop-gap. The only problem there is those decisions might hold us up a little longer if we try to get it super cleaned up.

File: 0014-tina-config-ts.md (hunk 1 above)
Comment (jeffsee55) [reply 0014-tina-config-ts.md:68]: Yeah, in theory we could import these into the runtime code but I'm not sure how much friction that would create.

**Comment by jamespohalloran** [comment]
> @jamespohalloran yeah I think we could go ahead and remove Forestry-isms in this process. The compilation step would add them back under the hood, but just as a stop-gap. The only problem there is those decisions might hold us up a little longer if we try to get it super cleaned up.

Yeah, I don't want to bog down this RFC with those decisions so maybe we'll discuss those later. 
I think there are some very low hanging fruit things, such as removing the need for properties like: hide_body, admin_path, webhook_url, new_page_extension, etc.

Bigger changes like renaming `front_matter/template/x` -> `/models/x` might be more contentious, and should maybe be left for a separate PR

**Comment by jeffsee55** [comment]
Another piece of low-hanging fruit is to remove the confusion around `select` references. We could add a new `type` which is just called `reference`, it'd still transform to the old `yml` config for now

**Comment by DirtyF** [comment]
Note: 🙏 Could we please list the related **goals** to convey the impact we're striving for the user or the product?

1. Goals (Problem)
2. Background (Context)
3. Implementation (Solution)
4. Risks and Rabbit Holes
5. No Gos

**Comment by jamespohalloran** [comment]
@DirtyF, are you asking us to format RFC's in the style of Shape-up pitches (with those explicitly defined categories)? Or do you feel that we are missing some relative information here? (I'm only asking because I feel there's a decent level of goals, background, and no-go's defined here (especially for a first pass)).

**Comment by jeffsee55** [comment]
Shower thought 💡  - we don't __need__ the `type` field. Only if the `component` field is __not__ a Tina core field like `text`, `group` etc. This way most Tina forms are already valid configs

**Comment by dwalkr** [comment]
I support this approach. Configuration in actual programming languages are extremely powerful and popular for a reason. It just - as you said - makes things more challenging when the config has to be read by an external application, which is where a declarative, "dumb" language like JSON have an advantage. YAML seems to be the worst of both worlds 😅 

> Shower thought 💡 - we don't **need** the `type` field. Only if the `component` field is **not** a Tina core field like `text`, `group` etc. This way most Tina forms are already valid configs

The "core fields" are registered as plugins, and conceivably be replaced by different fields. Seems unlikely that this would change the shape of data, but if you want fields to carry an explicit `type`, that metadata could conceivably be included with the field plugin. 

This would additionally make it possible for users to dictate the types of their custom fields by registering them as plugins (which, if we need to serialize the form config, might already be a requirement?)

**Comment by DirtyF** [comment]
> @DirtyF, are you asking us to format RFC's in the style of Shape-up pitches (with those explicitly defined categories)? Or do you feel that we are missing some relative information here? (I'm only asking because I feel there's a decent level of goals, background, and no-go's defined here (especially for a first pass)).

@jamespohalloran see https://github.com/vercel/next.js/discussions/20030 for how Vercel does it. Starting with explicit goals is great for product persons that are not interested in the implementation details.

**Comment by jeffsee55** [comment]
> The "core fields" are registered as plugins, and could conceivably be replaced by different fields. Seems unlikely that this would change the shape of data, but if you want fields to carry an explicit type, that metadata could conceivably be included with the field plugin.

Yeah, thinking about this more it just sounds like an extra `type` field is easiest. I was hoping there'd be a slick way to take an existing Tina config and just plop it in the `./.tina` folder, but I think that'd lead to more problems. So you _could_ give us an existing Tina form config, you just need that extra `type` field, and we'd ignore the `component` field for now. In the near future we can pass the `component` field through too, if provided.

It's also a good point of disconnect, while your `component` prop may implicitly return a `string`, for example - there's no guarantee that it will. Adding `type` is that guarantee - could be something we roll back into Tina open-source and act as a minimal built-in validation.

**Comment by jeffsee55** [comment]
I think this RFC should be scoped to immediate changes we can make for the next release. But this does lead into the discussion around "primitive" types, something @jpatters and @jamespohalloran have talked about in the past, where the `type` is just a shape, and nothing more, so the backend only cares about the shape, and the frontend is more concerned with how the field is presented.

So `image` and `text` are both actually just a `string`, and `group-list` is actually just `list` + `group` combo, and `tags` are just `list` + `string` combo. This really reduces the amount of concepts for the backend to know about - and probably opens the door for better extensibility. I want to put a pin in that thought for now, but just to acknowledge that this feels like the direction we're going and I think it's a strong concept we should really start to think about.

**Comment by jamespohalloran** [comment]
> I think this RFC should be scoped to immediate changes we can make for the next release. 

Yeah good call, I may have started a slippery slope :)

**Comment by jeffsee55** [comment]
We're going to move forward with this, a rough draft will be ready by Friday to discuss

**Pull request closed without merging**